### PR TITLE
Describe async iterators

### LINKED ISSUE #23410: The documentation of the iterators concept is incomplete
This page does not mention that an iterator method can be a source of not only `IEnumerable<T>`s, but also of `IAsyncEnumerable<T>`s.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: e8ba3a4c-1796-84b8-4152-b9806483ce86
* Version Independent ID: bd878489-9a4c-6d99-6c4c-c56e0387012b
* Content: [Iterators](https://docs.microsoft.com/en-us/dotnet/csharp/iterators)
* Content Source: [docs/csharp/iterators.md](https://github.com/dotnet/docs/blob/main/docs/csharp/iterators.md)
* Product: **dotnet-csharp**
* Technology: **csharp-advanced-concepts**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #23410

This article did not describe async iterators.

### CODE REVIEW

**Comment by BillWagner** [comment]
This will be much easier to review commit-by-commit.  

- The first commit is a general edit pass, and updating the samples to use extracted snippets validated by our CI system.
- The second commit updates the article to include a description of async enumerators.

File: docs/csharp/iterators.md [hunk 1]
```diff
@@ -144,117 +78,65 @@ public IEnumerable<int> GetSingleDigitNumbers()
...
-}
-```
+:::code language="csharp" source="./snippets/iterators/Generators.cs" ID="ReturnAndYieldReturn":::
 
-Look at the methods above. The first uses the standard `return` statement to return
-either an empty collection, or the iterator created by the second method. The second
-method uses the `yield return` statement to create the requested sequence.
+Look at the methods above. The first uses the standard `return` statement to return either an empty collection, or the iterator created by the second method. The second method uses the `yield return` statement to create the requested sequence.
 
 ## Deeper Dive into `foreach`
```
Comment (IEvangelist) [inline docs/csharp/iterators.md:93]:
```suggestion
## Deeper dive into `foreach`
```

File: docs/csharp/iterators.md [hunk 2]
```diff
@@ -144,117 +78,65 @@ public IEnumerable<int> GetSingleDigitNumbers()
...
-    Console.WriteLine(item.ToString());
+    var enumerator = collection.GetEnumerator();
+    try
+    {
+        while (enumerator.MoveNext())
+        {
+            var item = enumerator.Current;
+            Console.WriteLine(item.ToString());
+        }
+    } finally
```
Comment (IEvangelist) [inline docs/csharp/iterators.md:113]:
```suggestion
    }
    finally
```

File: docs/csharp/iterators.md [hunk 3]
```diff
@@ -144,117 +78,65 @@ public IEnumerable<int> GetSingleDigitNumbers()
...
     var enumerator = collection.GetEnumerator();
     try
     {
-        while (enumerator.MoveNext())
+        while (await enumerator.MoveNext())
         {
             var item = enumerator.Current;
             Console.WriteLine(item.ToString());
         }
     } finally
```
Comment (IEvangelist) [inline docs/csharp/iterators.md:132]:
```suggestion
    }
    finally
```

**Review by IEvangelist — APPROVED** [review]
Looks great - I only have a few nits.

**Pull request merged**